### PR TITLE
Add an automatic GCP-BOM dependency upgrader

### PR DESCRIPTION
**Pull request by Abacn**
Command: `python scripts/tools/gcpbomupgrader.py 26.30.0`

If run `python scripts/tools/gcpbomupgrader.py 26.31.0` based on this PR, it would give:

```
INFO:root:Changed com.google.api:gax: 2.39.0 -> 2.41.0
INFO:root:Changed com.google.cloud:google-cloud-spanner: 6.56.0 -> 6.57.0
INFO:root:Changed io.grpc:grpc-netty: 1.60.0 -> 1.61.0
INFO:root:Unchanged: io.netty:netty-transport:4.1.100.Final
INFO:root:Changed com.google.protobuf:protobuf-java: 3.25.1 -> 3.25.2
INFO:root:Changed org.apache.arrow:arrow-memory-core: 14.0.2 -> 15.0.0
```

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by Abacn** [comment]
CC: @dhruvdua @shunping

File: scripts/tools/gcpbomupgrader.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+      fout.write('\n')
+
+  def resolve(self):
+    logging.info("-----resolve dependency-----")
+    subp = subprocess.run([
+        self.runnable,
+        *('-q dependencies --configuration implementation --console=plain'
+          .split())
+    ],
+                          cwd=self.BUILD_DIR,
```
Comment (shunping) [inline scripts/tools/gcpbomupgrader.py:136]: nits: the format looks strange here

File: scripts/tools/gcpbomupgrader.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+Specifically, it
+
+1. preprocessing BeamModulePlugin.groovy to decide the dependencies need to sync
+2. generate an empty Maven project to fetch the exact target versions to change
+3. Write back to BeamModulePlugin.groovy
+
+There are few reasons we need to declare the version numbers:
+1. Sync the dependency that not included in GCP-BOM with those included with BOM
+  For example, "com.google.cloud:google-cloud-spanner" does while "com.google.cloud:google-cloud-spanner:():test" doesn't
+2. There are Beam artifacts not depending on GCP-BOM but used dependency managed
```
Comment (shunping) [inline scripts/tools/gcpbomupgrader.py:34]: Could you also document what we should change BeamModulePlugin if we have a new dep that we want to pin and we want to leverage this tool later (or at least don't break the standardization)?

There are two cases here:
1) adding a new dep from a project that is already tracked in BeamModulePlugin, and 2) adding a new dep from a new project.

**Review by shunping — COMMENTED** [review]
That's a great tool to facilitate libraries-bom version upgrade.

File: scripts/tools/gcpbomupgrader.py (hunk 2 above)
Comment (Abacn) [reply scripts/tools/gcpbomupgrader.py:34]: Good point. Definitely will clean up this script when marked as ready for review

Comment (Abacn) [reply scripts/tools/gcpbomupgrader.py]: Added comments about the tags that the script relies on: https://github.com/apache/beam/pull/30262/files#diff-0435a83a413ec063bf7e682cadcd56776cd18fc878f197cc99a65fc231ef2047R595

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 3]
```diff
@@ -601,17 +604,17 @@ class BeamModulePlugin implements Plugin<Project> {
...
     def google_cloud_bigdataoss_version = "2.2.16"
-    // Try to keep google_cloud_spanner_version consistent with google_cloud_spanner_bom in google_cloud_platform_libraries_bom
+    // [bomupgrader] determined by: com.google.cloud:google-cloud-spanner, consistent with: google_cloud_platform_libraries_bom
     def google_cloud_spanner_version = "6.57.0"
     def google_code_gson_version = "2.10.1"
     def google_oauth_clients_version = "1.34.1"
-    // Try to keep grpc_version consistent with gRPC version in google_cloud_platform_libraries_bom
-    def grpc_version = "1.60.0"
+    // [bomupgrader] determined by: io.grpc:grpc-netty, consistent with: google_cloud_platform_libraries_bom
+    def grpc_version = "1.61.0"
```
Comment (Abacn) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:617]: we again have the grpc version mismatch after #30181, my bad, fixed by this script

**Comment by shunping** [comment]
The code looks good. I am also wondering if we can use your script to add a presubmit test, so if the version is not right, the update cannot be submitted.

This may need to change your script to support two modes: in-place updating (which is already implemented) and reporting only. Then a new test can be added to call this script in reporting mode and check if there is any version mismatch.

WDYT?

**Comment by Abacn** [comment]
> The code looks good. I am also wondering if we can use your script to add a presubmit test, so if the version is not right, the update cannot be submitted.
> 
> This may need to change your script to support two modes: in-place updating (which is already implemented) and reporting only. Then a new test can be added to call this script in reporting mode and check if there is any version mismatch.
> 
> WDYT?

This sounds good, similar to .github/workflows/update_python_dependencies.yml . We can setup a "test" to generate a PR / or fail like the referred workflow.

However in practice we already have many tests and there are infra related workflow no one cares and red for months. So the actual benefit I am not sure. As of the scope of this PR I am not intend to setup a test.

**Comment by shunping** [comment]
LGTM!

**Pull request merged**